lib/pluginStorage.ts

```diff
@@ -0,0 +1,420 @@
+  deleteAll(locationType?: StorageLocationType): Promise<void>;
+
+  getSync(key: string, locationType?: StorageLocationType): any;
+  setSync(key: string, value: any, locationType?: StorageLocationType): void;
+  deleteSync(key: string, locationType?: StorageLocationType): void;
+  getAllSync(locationType?: StorageLocationType): { [key: string]: any };
+  setAllSync(dict: { [key: string]: any }, locationType?: StorageLocationType): void;
+  deleteAllSync(locationType?: StorageLocationType): void;
+}
+
+export interface IPluginSyncStorage {
+  get(key: string, locationType?: StorageLocationType): any;
+  set(key: string, value: any, locationType?: StorageLocationType): void;
+  delete(key: string, locationType?: StorageLocationType): void;
+  getAll(locationType?: StorageLocationType): { [key: string]: any };
+  setAll(dict: { [key: string]: any }, locationType?: StorageLocationType): void;
+  deleteAll(locationType?: StorageLocationType): void;
+
+}
+
+export interface ILocalStorage {
+  get(key: string, pluginId?: string): any;
+  set(key: string, value: any, pluginId?: string): void;
+  delete(key: string, pluginId?: string): void;
+  getAll(pluginId?: string): any;
+  setAll(dict: { [key: string]: any }, pluginId?: string): void;
+}
+
+export interface IHAStorage {
+  get(key: string, locationType?: StorageLocationType): Promise<any>;
+  set(key: string, value: any, locationType?: StorageLocationType): Promise<void>;
+  delete(key: string, locationType?: StorageLocationType): Promise<void>;
+  getAll(locationType?: StorageLocationType): Promise<any>;
+  setAll(dict: { [key: string]: any }, locationType?: StorageLocationType): Promise<void>;
+  deleteAll(locationType?: StorageLocationType): Promise<void>;
+}
+
+export function PluginStorageSyncFactory(pluginId: string, logger): IPluginSyncStorage {
+  const storage:any = PluginStorageFactory(pluginId, logger);
+  storage.get = storage.getSync;
+  storage.set = storage.setSync;
+  storage.delete = storage.deleteSync;
+  storage.getAll = storage.getAllSync;
+  storage.setAll = storage.setAllSync;
+  storage.deleteAll = storage.deleteAllSync;
+  return (storage as IPluginSyncStorage);
+}
+
+export function PluginStorageFactory(pluginId: string, logger): IPluginStorage {
+
+  const localStorage: ILocalStorage = LocalStorageFactory(pluginId);
+  const haStorage = apimlStorage.isConfigured() ? apimlStorage.makeStorageForPlugin(pluginId) : undefined;
+
+  if (process.clusterManager) { // Cluster mode
+    process.clusterManager.getStorageAll(pluginId).then(() => {
+      /* Then, once the cluster is done creating its own storage, we merge the local one
+         with the cluster. This is useful for having storage capability even if clusterManager hasn't
+         fully finished yet i. e. at startup, inside an app's constructor */
+      process.clusterManager.mergeStorage(pluginId, localStorage.getAll(pluginId));
+    });
+  }
+
+  return {
```

Comment:
the code is organized this way to try to keep the pluginId isolated to make it harder for plugins to interfere with each other.